pkg/object/cifs.go

```diff
@@ -0,0 +1,494 @@
+		limit: -1,
+	}, nil
+}
+
+// smbReadCloser is a wrapper for SMB file that handles connection release on close
+type smbReadCloser struct {
+	f     *smb2.File
+	conn  *cifsConn
+	store *cifsStore
+	limit int64
+	read  int64
+}
+
+func (r *smbReadCloser) Read(p []byte) (n int, err error) {
+	if r.limit >= 0 {
+		if r.read >= r.limit {
+			return 0, io.EOF
+		}
+		if int64(len(p)) > r.limit-r.read {
+			p = p[:r.limit-r.read]
+		}
+	}
+	n, err = r.f.Read(p)
+	r.read += int64(n)
+	return
+}
+
+func (r *smbReadCloser) Close() error {
+	err := r.f.Close()
+	r.store.releaseConnection(r.conn, err)
+	return err
+}
+
+func (c *cifsStore) Put(key string, in io.Reader, getters ...AttrGetter) (err error) {
+	return c.withConn(func(conn *cifsConn) error {
+		p := c.path(key)
+		if strings.HasSuffix(p, dirSuffix) {
+			// perm will not take effect, is not used
+			// ref: https://github.com/hirochachacha/go-smb2/blob/c8e61c7a5fa7bcd1143359f071f9425a9f4dda3f/client.go#L341-L370
+			return conn.share.MkdirAll(p, 0755)
+		}
+
+		// Ensure parent directories exist
+		dirPath := path.Dir(p)
+		if dirPath != "/" {
+			err = conn.share.MkdirAll(dirPath, 0755)
+			if err != nil {
+				return err
+			}
+		}
+
+		var tmp string
+		if PutInplace {
+			tmp = p
+		} else {
+			name := path.Base(p)
+			if len(name) > 200 {
+				name = name[:200]
+			}
+			tmp = path.Join(path.Dir(p), fmt.Sprintf(".%s.tmp.%d", name, rand.Int()))
+			defer func() {
+				if err != nil {
+					_ = conn.share.Remove(tmp)
+				}
+			}()
+		}
+
+		f, err := conn.share.Create(tmp)
+		if err != nil {
+			return err
+		}
+
+		buf := bufPool.Get().(*[]byte)
+		defer bufPool.Put(buf)
+		_, err = io.CopyBuffer(f, in, *buf)
+		if err != nil {
+			_ = f.Close()
+			return err
+		}
+
+		err = f.Close()
+		if err != nil {
+			return err
+		}
+
+		if !PutInplace {
+			_ = conn.share.Remove(p) // Ignore error if file doesn't exist
+			return conn.share.Rename(tmp, p)
+		}
+		return nil
+	})
+}
+
+func (c *cifsStore) Delete(key string, getters ...AttrGetter) (err error) {
+	return c.withConn(func(conn *cifsConn) error {
+		p := strings.TrimRight(c.path(key), dirSuffix)
+		err = conn.share.RemoveAll(p)
+		if err != nil && os.IsNotExist(err) {
+			err = nil
+		}
+		return err
+	})
+}
+
+func (c *cifsStore) fileInfo(key string, fi os.FileInfo, isSymlink bool) Object {
+	owner, group := "nobody", "nobody"
+	ff := &file{
+		obj{key, fi.Size(), fi.ModTime(), fi.IsDir(), ""},
+		owner,
+		group,
+		fi.Mode(),
+		isSymlink,
+	}
+	if fi.IsDir() {
+		if key != "" && !strings.HasSuffix(key, "/") {
+			ff.key += "/"
+		}
+	}
+	return ff
+}
+
+func (c *cifsStore) List(prefix, marker, token, delimiter string, limit int64, followLink bool) ([]Object, bool, string, error) {
+	if delimiter != "/" {
+		return nil, false, "", notSupported
+	}
+
+	dir := c.path(prefix)
+	if !strings.HasSuffix(dir, "/") {
+		dir = path.Dir(dir)
+		if !strings.HasSuffix(dir, dirSuffix) {
+			dir += dirSuffix
+		}
+	}
+	var mEntries []*mEntry
```

Comment:
If marker is empty, need HEAD prefix ref [file.go](https://github.com/juicedata/juicefs/blob/779801773ff153cddbb38d0f2af146e1993a79f3/pkg/object/file.go#L291-L291)